In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
pd.set_option('display.max_columns', None)

sns.set_style('whitegrid')

In [3]:
df = pd.read_excel('../data/raw/raw_data.xlsx')

In [4]:
df.head()

,SKU_ID,Category,ABC_Class,Supplier_Name,Warehouse_Location,Received_Date,Last_Purchase_Date,Stock_Age_Days,Quantity_On_Hand,Damaged_Qty,Returns_Qty,Avg_Daily_Sales,Forecast_Next_30d,Reorder_Point,Safety_Stock,Lead_Time_Days,Total_Inventory_Value_USD,Supplier_OnTime_Pct,Inventory_Status
0,SKU_14996,Household,A,NextGen Traders,Chennai Warehouse,2024-06-11,2024-07-11,212,482.0,13,9,19.84,459.32,126,47,7.0,45555.63,82.64,Normal
1,SKU_10247,Snacks,C,NextGen Traders,Bangalore Warehouse,2024-11-24,2025-03-19,357,640.0,8,6,34.81,1072.54,110,75,5.0,36577.48,94.58,Normal
2,SKU_16215,Dairy,C,NextGen Traders,Chennai Warehouse,2023-02-01,2023-03-06,104,538.0,5,3,49.66,1891.81,118,66,12.0,23742.67,98.22,Overstock
3,SKU_13545,Bakery,A,Prime Distribution,Chennai Warehouse,2024-03-13,2024-06-04,161,530.0,9,2,20.00,662.42,119,11,8.0,49287.57,97.45,Low Stock
4,SKU_15397,Produce,B,Global Retail Supply,Chennai Warehouse,2023-05-01,2023-05-09,289,397.0,14,3,32.63,1115.41,128,42,3.0,16929.86,86.75,Overstock


### Observation
- Dataset contains inventory operational data.
- Includes supplier, warehouse, sales, and stock information.
- Date columns appear available for time-based analysis.

In [5]:
df.shape

(6712, 19)

### Observation
- Dataset contains 6712 rows and 19 columns.
- Sufficient for exploratory data analysis.

In [6]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 6712 entries, 0 to 6711
Data columns (total 19 columns):
 #   Column                     Non-Null Count  Dtype         
---  ------                     --------------  -----         
 0   SKU_ID                     6712 non-null   str           
 1   Category                   6712 non-null   str           
 2   ABC_Class                  6712 non-null   str           
 3   Supplier_Name              6712 non-null   str           
 4   Warehouse_Location         6712 non-null   str           
 5   Received_Date              6511 non-null   datetime64[us]
 6   Last_Purchase_Date         6712 non-null   datetime64[us]
 7   Stock_Age_Days             6712 non-null   int64         
 8   Quantity_On_Hand           6712 non-null   float64       
 9   Damaged_Qty                6712 non-null   int64         
 10  Returns_Qty                6712 non-null   int64         
 11  Avg_Daily_Sales            6712 non-null   float64       
 12  Forecast_Next_30d

### Observation
- Dataset contains numerical and categorical variables.
- Some columns contain missing values.
- Date columns may require datatype conversion.

In [7]:
df.describe()

,Received_Date,Last_Purchase_Date,Stock_Age_Days,Quantity_On_Hand,Damaged_Qty,Returns_Qty,Avg_Daily_Sales,Forecast_Next_30d,Reorder_Point,Safety_Stock,Lead_Time_Days,Total_Inventory_Value_USD,Supplier_OnTime_Pct
count,6511,6712,6712.000000,6712.000000,6712.000000,6712.000000,6712.000000,6509.000000,6712.000000,6712.000000,6510.000000,6712.000000,6510.000000
mean,2024-02-28 12:18:01.492858,2024-04-29 20:31:02.216925,203.411651,420.498784,9.602354,7.546633,34.889443,1044.936906,119.042312,59.629917,9.554685,28970.025240,90.547671
min,2023-01-01 00:00:00,2023-01-08 00:00:00,5.000000,5.000000,0.000000,0.000000,1.000000,20.840000,10.000000,5.000000,1.000000,251.060000,82.000000
25%,2023-07-28 00:00:00,2023-09-27 00:00:00,103.000000,339.000000,6.000000,5.000000,28.080000,779.400000,98.000000,46.000000,8.000000,16374.687500,86.380000
50%,2024-02-29 00:00:00,2024-04-27 00:00:00,205.000000,421.000000,10.000000,8.000000,35.000000,1017.630000,119.000000,60.000000,10.000000,27061.900000,90.520000
75%,2024-09-30 00:00:00,2024-12-02 00:00:00,302.000000,500.000000,13.000000,10.000000,41.750000,1285.400000,140.000000,73.000000,12.000000,39306.775000,94.760000
max,2025-04-30 00:00:00,2025-08-25 00:00:00,400.000000,1320.229621,28.000000,25.000000,70.290000,2478.370000,237.000000,127.000000,21.000000,86760.458785,98.990000
std,NaN,NaN,114.147916,122.066614,4.865126,3.914893,10.028864,366.038167,30.194679,19.980234,3.001102,15264.646666,4.886313


### Observation
- Large variation exists in inventory quantities and values.
- Some features may contain extreme values.

In [8]:
df.describe(include=['object', 'string'])

,SKU_ID,Category,ABC_Class,Supplier_Name,Warehouse_Location,Inventory_Status
count,6712,6712,6712,6712,6712,6712
unique,6700,8,3,6,5,4
top,SKU_15166,Produce,A,Prime Distribution,Kolkata Warehouse,Normal
freq,2,869,2296,1133,1372,1761


In [9]:
df.nunique().sort_values(ascending=False)

SKU_ID                       6700
Total_Inventory_Value_USD    6685
Forecast_Next_30d            6338
Avg_Daily_Sales              3249
Supplier_OnTime_Pct          1664
Last_Purchase_Date            941
Received_Date                 851
Quantity_On_Hand              635
Stock_Age_Days                396
Reorder_Point                 194
Safety_Stock                  120
Damaged_Qty                    28
Returns_Qty                    23
Lead_Time_Days                 21
Category                        8
Supplier_Name                   6
Warehouse_Location              5
Inventory_Status                4
ABC_Class                       3
dtype: int64

In [10]:
missing = df.isnull().sum()

missing_percent = (missing / len(df)) * 100

missing_df = pd.DataFrame({
    'Missing Values': missing,
    'Percentage': missing_percent
})

missing_df.sort_values(
    by='Percentage',
    ascending=False
)

,Missing Values,Percentage
Forecast_Next_30d,203,3.024434
Lead_Time_Days,202,3.009535
Supplier_OnTime_Pct,202,3.009535
Received_Date,201,2.994636
ABC_Class,0,0.000000
SKU_ID,0,0.000000
Category,0,0.000000
Stock_Age_Days,0,0.000000
Supplier_Name,0,0.000000
Warehouse_Location,0,0.000000


In [11]:
df.duplicated().sum()

np.int64(12)

In [12]:
num_cols = df.select_dtypes(include=np.number).columns.tolist()

cat_cols = df.select_dtypes(include='object').columns.tolist()

print("Numerical Columns:")
print(num_cols)

print("\nCategorical Columns:")
print(cat_cols)

Numerical Columns:
['Stock_Age_Days', 'Quantity_On_Hand', 'Damaged_Qty', 'Returns_Qty', 'Avg_Daily_Sales', 'Forecast_Next_30d', 'Reorder_Point', 'Safety_Stock', 'Lead_Time_Days', 'Total_Inventory_Value_USD', 'Supplier_OnTime_Pct']

Categorical Columns:
['SKU_ID', 'Category', 'ABC_Class', 'Supplier_Name', 'Warehouse_Location', 'Inventory_Status']


C:\Users\ASUS\AppData\Local\Temp\ipykernel_7832\2097288367.py:3: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  cat_cols = df.select_dtypes(include='object').columns.tolist()
